### PR TITLE
Host doxygen on ReadtheDocs

### PR DESCRIPTION
**Pull request by bmhan12**
- This PR gets ReadtheDocs to host doxygen documentation. 
You can view the doxygen by going to this branch's [ReadTheDocs](https://axom.readthedocs.io/en/docs-han12-doxygenonreadthedocs/) and clicking on one of the links under "Source Code Documentation" on the first page.
 
- I did a pass through and fixed any broken links I found in reStructuredText and README.md files (README's get pulled into the doxygen documentation).  There were a lot, so I probably did miss some… I also updated links in "References and Useful Resources": https://axom.readthedocs.io/en/docs-han12-doxygenonreadthedocs/docs/sphinx/coding_guide/references.html
so please let me know if they do not link to where they are supposed to.
 
- Added a path to sphinx-build (SPHINX_EXECUTABLE) in rzgenie-toss_3_x86_64_ib-gcc@8.1.0.cmake .  Motivation being I got an error when trying to generate sphinx documentation with the version at /usr/bin/sphinx-build (version 1.1.3)

### CODE REVIEW

File: src/docs/doxygen/CMakeLists.txt [hunk 1]
```diff
@@ -9,5 +9,10 @@
...
 blt_add_doxygen_target( axom_doxygen )
 
+# Required directory for doxygen to generate output to 
+add_custom_target( build_doxygen_output_dir 
+                   ALL 
+                   COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_BINARY_DIR}/html 
+                 ) 
 
-
+add_dependencies( axom_doxygen build_doxygen_output_dir )
```
Comment (gzagaris) [inline src/docs/doxygen/CMakeLists.txt]: add newline to the end of file

File: src/conf.py [hunk 2]
```diff
@@ -16,6 +16,33 @@
...
+    cwd=os.getcwd()
+    buildpath=os.path.join(cwd,"_build")
+    if (os.path.isdir(buildpath) == 0):
+        os.mkdir(buildpath)
+    htmlpath=os.path.join(buildpath,"html")
+    if (os.path.isdir(htmlpath) == 0):
+        os.mkdir(htmlpath)
+
+    # Modify Doxyfile for ReadTheDocs compatibility
+    with open('./docs/doxygen/Doxyfile.in', 'r') as f:
```
Comment (gzagaris) [inline src/conf.py]: Could we enable the `ALPHABETICAL_INDEX` setting in our Doxyfile? I tend to use that the most to navigate the documentation.

File: src/axom/lumberjack/README.md [hunk 3]
```diff
@@ -3,5 +3,5 @@ Lumberjack: Log filtering and aggregation {#lumberjacktop}
 
 [Lumberjack](@ref axom::lumberjack) provides functionality to filter log events based on priority, rate, and volume.  It coordinates with but does not require the [SLIC](@ref slictop) component.
 
-The [introduction to Lumberjack](../../../sphinx/axom_docs/html/axom/lumberjack/docs/sphinx/index.html)
+The [introduction to Lumberjack](https://axom.readthedocs.io/en/develop/axom/lumberjack/docs/sphinx/index.html)
```
[the remark this answers is not in the document]
Comment (kennyweiss) [reply src/axom/lumberjack/README.md]: I just checked and these files are used in two ways:
* They are the ``README`` files on github for the component directories, 
   e.g. https://github.com/LLNL/axom/blob/0ae4f5dc8d432d9f93f5e9f58922c80c8b5c00dc/src/axom/lumberjack/README.md
* They are the landing page in doxygen for the given component, 
  e.g. https://axom.readthedocs.io/en/docs-han12-doxygenonreadthedocs/doxygen/html/lumberjacktop.html

In both cases, this PR is hard-coding them to point to the ``develop`` branch on read the docs.

File: src/docs/doxygen/CMakeLists.txt (hunk 1 above)
Comment (bmhan12) [reply src/docs/doxygen/CMakeLists.txt]: done

File: src/conf.py (hunk 2 above)
Comment (bmhan12) [reply src/conf.py]: That is really convenient! Added into Doxyfile.in

File: host-configs/rzgenie-toss_3_x86_64_ib-gcc@8.1.0.cmake [hunk 4]
```diff
@@ -58,6 +58,7 @@ set(UMPIRE_DIR "${TPL_ROOT}/umpire-1.0.0/share/umpire/cmake" CACHE PATH "")
 # python not built by uberenv
 
 set(ENABLE_DOCS OFF CACHE BOOL "")
+set(SPHINX_EXECUTABLE /collab/usr/gapps/python/build/spack-toss3.2/opt/spack/linux-rhel7-x86_64/gcc-4.9.3/python-2.7.14-7rci3jkmuht2uiwp433afigveuf4ocnu/bin/sphinx-build CACHE PATH "")
```
[the remark this answers is not in the document]
Comment (bmhan12) [reply host-configs/rzgenie-toss_3_x86_64_ib-gcc@8.1.0.cmake]: @kennyweiss `make docs` is a valid target in that case, but nothing gets generated unless `ENABLE_DOCS` is `ON`.

File: src/axom/lumberjack/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bmhan12) [reply src/axom/lumberjack/README.md]: @kennyweiss I agree, it is not ideal for the links to be hard coded to the develop branch. I tried relative paths for these `README.md` pages, which works for the` doxygen` landing pages, but are broken links on GitHub.  Hard-coded links work are non-broken in both cases, but not great...

File: src/docs/doxygen/Doxyfile.in [hunk 5]
```diff
@@ -58,7 +58,7 @@ PROJECT_LOGO           =
 # entered, it will be relative to the location where doxygen was started. If
 # left blank the current directory will be used.
 
-OUTPUT_DIRECTORY       =
+OUTPUT_DIRECTORY       = ../../html/doxygen
```
[the remark this answers is not in the document]
Comment (bmhan12) [reply src/docs/doxygen/Doxyfile.in]: @kennyweiss Yes, this does work for local builds.  For example, if you call `make docs` inside a build directory, then the generated html docs will be located at `build_dir/html`, with the generated doxygen located at `build_dir/html/doxygen/html`. I believe the `doxygen` process is run in a subdirectory of the build directory (wherever the `Doxyfile` is located), which for me is `build_dir/docs/doxygen/Doxyfile`.

[the remark this answers is not in the document]
Comment (bmhan12) [reply src/docs/doxygen/Doxyfile.in]: @kennyweiss I changed the `OUTPUT_DIRECTORY` to be an absolute path to make things a little more clear.

**Comment by bmhan12** [comment]
@kennyweiss  All `develop` path links have been removed from the `README's`, as well as some hard-coded `develop` links I missed the first time in the `reStructuredText` files.

**Comment by kennyweiss** [comment]
Thanks @bmhan12 -- this is great!

File: src/axom/mint/docs/sphinx/sections/citations.rst [hunk 6]
```diff
@@ -6,13 +6,13 @@
 
 .. _LLNL: https://www.llnl.gov/
 .. _Umpire: https://umpire.readthedocs.io/en/develop/
-.. _Mint Doxygen API Documentation: https://lc.llnl.gov/axom/docs/doxygen/axom_doxygen/html/classaxom_1_1mint_1_1Mesh.html
-.. _Axom Quick Start Guide: https://axom.readthedocs.io/en/develop/docs/sphinx/quickstart_guide/index.html
+.. _Mint Doxygen API Documentation: ../../../../../doxygen/html/minttop.html
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/citations.rst]: Building the documentation gives complaints "duplicate explicit targets" for lines 9, 10, 14, and 15.  Does this file actually get linked to?  @gzagaris , could we delete this file?
Edit: See also slic's file citations.rst.

File: src/axom/mint/docs/sphinx/sections/synopsis.rst [hunk 7]
```diff
@@ -40,12 +40,12 @@ the following third-party libraries is provided:
 
 * `RAJA`_, used for the parallel execution and portability layer.
 
-* `Conduit`_ , for using `Sidre`_ as the :ref:`MeshStorageManagement` system.
+* `Conduit`_ , for using `Sidre <../../../sidre/docs/sphinx/index.html>`_ as the :ref:`MeshStorageManagement` system.
```
Comment (agcapps) [inline src/axom/mint/docs/sphinx/sections/synopsis.rst]: I think this file (mint's sphinx synopsis.rst) was replaced by index.rst.  It doesn't look like synopsis.rst is linked to from anywhere, and should be removed.  I realized this when I looked at the warnings that came up when this file was built.  Thoughts, @bmhan12 and @gzagaris ?

File: src/axom/slic/docs/sphinx/sections/citations.rst [hunk 8]
```diff
@@ -5,10 +5,10 @@
 
 .. _LLNL: https://www.llnl.gov/
 .. _Shroud: https://shroud.readthedocs.io/en/develop/
-.. _Lumberjack: https://lc.llnl.gov/axom/docs/sphinx/lumberjack_docs/html/index.html
-.. _Axom Quick Start Guide: https://axom.readthedocs.io/en/develop/docs/sphinx/quickstart_guide/index.html
-.. _Axom Toolkit: https://axom.readthedocs.io/en/develop/
+.. _Lumberjack: ../../../../lumberjack/docs/sphinx/index.html
```
Comment (agcapps) [inline src/axom/slic/docs/sphinx/sections/citations.rst]: Sphinx complains about this line and the next two.  I suspect this file needs to be removed.  See also mint's citations.rst.

File: src/axom/lumberjack/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (agcapps) [reply src/axom/lumberjack/README.md]: Hmm, I just got to this today as part of my review and didn't realize you had already discussed these links.  Sorry about that.  It does seem that the least annoying/misleading thing to do is to take out the lines.

File: src/axom/lumberjack/README.md [hunk 9]
```diff
@@ -2,6 +2,3 @@ Lumberjack: Log filtering and aggregation {#lumberjacktop}
 ========
 
 [Lumberjack](@ref axom::lumberjack) provides functionality to filter log events based on priority, rate, and volume.  It coordinates with but does not require the [SLIC](@ref slictop) component.
-
-The [introduction to Lumberjack](../../../sphinx/axom_docs/html/axom/lumberjack/docs/sphinx/index.html)
```
[the remark this answers is not in the document]
Comment (agcapps) [reply src/axom/lumberjack/README.md]: I just saw your earlier conversation about these lines in all the README.md files.  Sorry about that.

[the remark this answers is not in the document]
Comment (bmhan12) [reply src/axom/lumberjack/README.md]: @agcapps No problem!

File: src/axom/mint/docs/sphinx/sections/synopsis.rst (hunk 7 above)
Comment (bmhan12) [reply src/axom/mint/docs/sphinx/sections/synopsis.rst]: @agcapps Based off the [raw text](https://raw.githubusercontent.com/LLNL/axom/develop/src/axom/mint/docs/sphinx/index.rst) for mint's `index.rst`, `index.rst` links to `synopsis.rst`. It's not visible for some reason in Github's [preview](https://github.com/LLNL/axom/blob/develop/src/axom/mint/docs/sphinx/index.rst) view of the file.

**Comment by bmhan12** [comment]
@agcapps I was able to remove the doxygen warnings by removing links to `citation.rst` from the  `index.rst` files for slic and mint, then adding the necessary custom links in `synopsis.rst`

File: src/axom/mint/docs/sphinx/sections/synopsis.rst (hunk 7 above)
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/synopsis.rst]: That's right `synopsis.rst` is included by the `index.rst` -- it consists of the short summary for each component that gets rendered on the component's landing page.

File: src/axom/mint/docs/sphinx/sections/citations.rst (hunk 6 above)
Comment (gzagaris) [reply src/axom/mint/docs/sphinx/sections/citations.rst]: We shouldn't delete this file. See comment below.

File: src/axom/mint/docs/sphinx/sections/synopsis.rst (hunk 7 above)
Comment (agcapps) [reply src/axom/mint/docs/sphinx/sections/synopsis.rst]: Aha!  Good reason not to delete the file.  Sorry I didn't see that immediately, George.

File: src/axom/slic/docs/sphinx/sections/citations.rst (hunk 8 above)
Comment (agcapps) [reply src/axom/slic/docs/sphinx/sections/citations.rst]: Sorry I didn't see this one was included, as is mint's citiation.rst.  Good reason not to delete it.

**Pull request merged**